### DISCUSSION TITLE
What to do about Tags

### DISCUSSION THREAD

**Discussion by hay-kot** [mealie-recipes/mealie#242] [Ideas]
Share your ideas on what you want Tags to do or how they should be utilized? I'm struggling to figure out exactly what role they play.

Questions I have
- What makes a category different than a tag?
- How to I want to interact with tags? 

Things I know will happen
- Search Page with Tag/Category Filter

**Comment by carlyman** [comment]
I like tags for things like: (1) "Grandma" or other family member for an old family recipe; (2) "Holiday" for favorite holiday recipes; (3) "Restaurant" for restaurant recipes, especially those shared during COVID; and (4) "Rick Bayless" or other chef such that I can see just their recipes.  I could see also having a Tag for the user (or user group) that created and user (or user group) that last modified the recipe -- this way if I share with my extended family, they could easily filter on their recipes.

Functionally once a recipe has a tag, I see interacting with them very similarly to Categories.  In the search box I should be able to search for a tag, when viewing a recipe I should be able to click the Tag name and then see a list of recipes with that tag, and when creating a Custom Page I should be able to use Tags and/or Categories.

I find Tags different than Categories which I use for Appetizer, Italian, Pasta, BBQ, Side Dish...essentially categories you'd find in a cook book.  In reality, I guess Categories and Tags could just be combined, but I sort of like having them separate.

**Comment by sephrat** [comment]
My categories are mostly based on the 'course' they belong to. Appetizers, Starters, Main course, but also sauces or beverages. When I'm looking for an idea for an every day dinner, I just browse "main course".
My tags are for picking the exact type of meal I'm looking for, based on my mood or my guests' diet, like gluten-free, vegetarian, sweet-sour or casserole. They can also act as sub-categories, like "alcool" for beverages or "hot meal" for a main course.

As for interacting, I'm thinking of 2 different ways:
- quick search: if I start typing a tag's name in the top right search bar, I get all recipes assigned to this tag.
- advanced search: a dedicated search page that allows to include/exclude categories and/or tags, so that I can search for any main course without the "shellfish" nor "cheese" tag for example.

**Comment by dietrichmd** [comment]
I use categories mainly for top-level things such as appetizers, drinks, breakfast, etc.  Tags for me are things like BEEF, Chicken, Vodka, Restaurant, etc.

I agree with @sephrat with regards to search.  Of course, quick search (the one up top on all pages) should include search-by-tag as part of its default searches, but that shouldn't need to be mentioned :)

**Comment by cysec404** [comment]
Just starting using Mealie and I'm a fan of the tags. I use categories like main course, sides, dessert, bread, and then I use tags for things like keto. But since you can add multiple categories to a recipe it might be redundant. Hmm.

**Comment by describe19** [comment]
As someone who just installed mealie and added a recipe, my very first thought was 'what is the difference between categories and tags'. Tags show up on the recipe card and categories don't - at the moment categories just seem like a worse version of tags. Is there difference between them when searching? As far as I can tell they are almost the same and should just be merged - they are both descriptors for a group of recipes. If you think there is reason to keep both, I'd like to hear.

**Reply by hay-kot** [reply]
I see categories as umbrellas and tags as more specific identifiers. As such, categories play a larger role in building pages and creating meals, I use categories sparingly. Tags on the other hand are for specific criteria like "vegan" or "gluten free". That way I can easily search the Desserts category for recipes that have the gluten free tag. Tags are shown on recipes instead of categories because in the discord it was brought up that if you're looking at a recipe, you probably know what category it's in, and showing what tags are attached is probably more useful. I find having both helps me more clearly define the scope of a recipe, but if you just want to use categories, I don't think there's anything stopping you from doing that.

**Comment by ltctceplrm** [comment]
I think solving the categories and tags problem would remove a lot of confusion and avoid bugs [like this one](https://github.com/hay-kot/mealie/issues/290). To keep it simple I'd just have categories be for the courses like several users have already mentioned: Dessert, Appetizer, Main Course,...
And tags for modifiers like vegan, gluten free, peanuts (in case anyone's alergic), mom's recipes,...
The categories would show up on the left side of the screen and if you select a category it takes you to the category page where you could then have a filter option at the top to filter out tags you don't want or actually do want for that category. Like it currently is on the search page (which is great by the way).

**Comment by ovizii** [comment]
I currently use them like this:

categories: breakfast, lunch, dinner, dessert
tags: Asian, Fast Food, Italian

not 100% happy with it though. Ideally there'd be more things in addition to categories and tags. 

How about being able to search and filter by:

category: Breakfast, Lunch, Dinner, Dessert
cuisine: Italian, Spanish, American, Fusion Food
ingredients: Pork, Potatoes,
tag: grandma, cookbook, tried & tested
preparation time: <30, <60, >60

If this would work I could quickly search for a Spanish recipe for Lunch with Pork which is tagged "tried & tested" with preparation time less <60 minutes.

**Reply by k9sartech** [reply]
The search function for tags includes multiple tags in a search.  Therefore in my understanding, each recipe could have a tag for each of those fields.  Category: Breakfast; Tag: American, Potatoes, Tried & Tested, <30 = Home Fries, Hash Browns.  You'd just have to remember to include a Cuisine, Main Ingredient, 'Tag', and Preparation Time tag in each recipe.  Tags can also include Preparation Method (Bake, Broil, Fry, Grill, Slow Cooker, Instapot, Air Fry, etc.), Season/Occasion (Christmas, Easter, Picnics, Superbowl, Winter, etc.), Special Considerations (Kid Friendly, Vegetarian, Vegan, Diabetic, Gluten-free. etc.).

**Comment by k9sartech** [comment]
Can Tags be renamed or deleted?  Same with Categories?  Can Categories be renamed rather then deleted and have to find recipes that were in the old category and put them in the new category?  I know Categories can be deleted.

**Reply by sephrat** [reply]
You cannot rename a category or a tag yet.
You can delete a category from the UI, and it seems that you can delete a tag through the API only (not tested).

**Reply by pto199** [reply]
How do I do this?  I'm setting up Mealie and made a few categories and tags I don't want.  I dont see any buttons to delete or rename tags&Catagories?

**Reply by reecespieces** [reply]
> How do I do this? I'm setting up Mealie and made a few categories and tags I don't want. I dont see any buttons to delete or rename tags&Catagories?

You can delete and rename both tags and categories. Under Settings -> Toolbox.

**Comment by GourmetSaint** [comment]
In the "Manage Data" area, is there a plan to be able to filter by column data?  When I imported a "Cookbook" from Paprika, it was tagged with "paprika". Now I want to filter all those with "paprika" tag and add a category or another tag? Any way I can do that? Anothercexample is, after migration from other recipe systems, some are tagged with "Beef" and others have "Beef Dishes". I want to search recipes by one tag and bulk update those recipe tags with the other before deleting the superfluous tag?

**Comment by lewisward1** [comment]
can someone help with updating a Tag via API >?

I tried this format in Insomnia but i get unknown error not really touched API before and cant find a complete example
{
  "id": "7714b117-9eac-4dfe-9bd3-1642e68035b2",
  "tags": [
    {
      "id": "123e4567-e89b-42d3-a456-426614174001",
      "name": "Sia",
      "slug": "sia",
      "group_id": "304e2b06-f947-4e4f-b99b-6d9cfef0d167"
    }
  ]
}

**Reply by Kuchenpirat** [reply]
you can visit the `/docs` page of your instance (e.g. [https://demo.mealie.io/docs](https://demo.mealie.io/docs)) for the demo instance. There you see all availlable api options. 

For updating an tag you are looking at this one: https://demo.mealie.io/docs#/Organizer%3A%20Tags/update_one_api_organizers_tags__item_id__put

This would be an example schema:
```json
{
  "name": "string",
  "id": "string",
  "slug": "string",
  "recipes": []
}
```

**Reply by lewisward1** [reply]
Thanks - something like this ?
{
  "name": "Sia",
  "id": "7714b117-9eac-4dfe-9bd3-1642e68035b2",
  "slug": "sia",
  "recipes": [test]
}



This is the reply

**Reply by lewisward1** [reply]
{
	"detail": [
		{
			"type": "json_invalid",
			"loc": [
				"body",
				101
			],
			"msg": "JSON decode error",
			"input": {},
			"ctx": {
				"error": "Expecting value"
			}
		}
	]
}

**Reply by Kuchenpirat** [reply]
easiest would be to update an tag through the gui. In the network tab of your browser you should be able to see what is send. 

Here i for example renamed the Tag American to American 2
path= `/api/organizers/tags/94039b52-4be6-496c-8e99-3647b02787d7`

Payload: 
```json 
{
    "id": "94039b52-4be6-496c-8e99-3647b02787d7",
    "name": "American2",
    "slug": "american"
}
```

**Reply by lewisward1** [reply]
ah ok im using the wrong one then - i wish to add a tag to multiple recipes - ive imported 120 and want to tag them to a cook book 
hence why i dont want to use the gui

**Reply by lewisward1** [reply]
network tab very useful - thanks

**Comment by ados8** [comment]
I've just started using Mealie and for me the tags are difficult to keep neat.
When you import a recipe you have 3 options regarding tags.

1. Assign no tags
2. Assign whatever tags it finds
3. Assign whatever tags it finds but manually correct tags with edit

The problem is even in the edit stage it floods the tag system making new tags even when it hasn't been saved yet.
This means I have to clean out the tag garbage after every import to keep the system from having 1,000 tags.
It would be really nice to have this tweaked, thoughts on new implementations which could all be added.

1. Assign no tags (unchanged)
2. Assign whatever tags it finds (unchanged)
3. Assign whatever tags it finds but manually correct tags with edit (until saved no ticked tags are created)
4. At the URL import screen a quick edit menu in addition the full new window option. Show identified tags, allow toggling them and setting the category which is the 2 most modified attributes for organisation. This would done without needing to use the main edit screen.
5. Assign tags it finds that match existing tags only (this would be a global setting in the household)

**Comment by fhp** [comment]
The way Linear handles tags would be a nice addition to Mealie. There you can make tag groups (like cuisine, season, etc. You can create your own groups). From each tag group, you can pick at most one tag.

<img width="672" height="447" alt="image" src="https://github.com/user-attachments/assets/3d4b39d8-ac44-4109-ade9-9ea100fcbccd" />

**Comment by fhp** [comment]
Another wishlist item would be private tags. My recipes are public, but I'd like to add some extra tags that are personal: 'favorite of Kid 1', 'favorite of Kid 2', but I don't want to broadcast the names and meal preferences of my kids on the internet :wink:
